src/utilities/utilities.jl

```diff
@@ -322,6 +322,67 @@ function surface_normals(x, y, z)
     return vec(map(normal, CartesianIndices(z)))
 end
 
+
+############################################################
+#                NaN-aware normal handling                 #
+############################################################
+
+"""
+    nan_aware_orthogonal_vector(v1, v2, v3) where N
+
+Returns an un-normalized normal vector for the triangle formed by the three input points.
+Skips any combination of the inputs for which any point has a NaN component.
+"""
+function nan_aware_orthogonal_vector(v1, v2, v3)
+    centroid = Vec3f(((v1 .+ v2 .+ v3) ./ 3)...)
+    normal = [0.0, 0.0, 0.0]
+    # if the coord is NaN, then do not add.
+    (isnan(v1) | isnan(v2)) || (normal += cross(v2 .- centroid, v1 .- centroid))
+    (isnan(v2) | isnan(v3)) || (normal += cross(v3 .- centroid, v2 .- centroid))
+    (isnan(v3) | isnan(v1)) || (normal += cross(v1 .- centroid, v3 .- centroid))
+    return Vec3f(normal).*-1
+end
```

Comment:
This produces a NaN vector with one NaN input because centroid cmbines all 3:
```julia
julia> Makie.nan_aware_orthogonal_vector(Point3f(0), Point3f(1,0,0), Point3f(NaN))
3-element Vec{3, Float32} with indices SOneTo(3):
 NaN
 NaN
 NaN
```
I'm not sure why this doesn't affect things down the line, but you can probably just do
```julia
any(isnan, (v1, v2, v3)) && return Vec3f(0)
return cross(v2 - v1, v3 - v1)
```
instead?

Reply:
`any(isnan, (v1, v2, v3))` this is in the drawing code, so I guess that's why it isn't a problem

Reply:
Ah, no, it is affecting things down the line.

If we have two faces
```
a --- b
| \   |
|   \ |
c --- d
```
and only b is NaN, orthogonal_vector will produce NaN as the face normal of (a, b, d). This is getting added to the normal at vertex a, b and d via
```julia
for i in 1:length(F)
    fi = face[i]
    normals_result[fi] = normals_result[fi] + n
end
```
When we iterate the faces, (a, b, d) but (a, c, d) is not. It seems like the shading is doing something differently though, so it's not as obvious.

---

With a plain color you can see the effect
```julia
zs = rand(10, 10)
ns = copy(zs)
ns[4, 3:6] .= NaN
f, a, p = surface(1..10, 1..10, ns, colormap = [:lightblue, :lightblue])
scatter!(a, [Point3f(i, j, zs[i, j]) for i in 1:10 for j in 1:10], color = isnan.(ns'[:]), depth_shift = -1f-3)
wireframe!(a, CairoMakie.surface2mesh(to_value.(p.converted)...), depth_shift = -1f-3, color = :black)
f
```

![Screenshot from 2023-01-17 16-13-12](https://user-images.githubusercontent.com/10947937/212936134-f2faa5c4-d251-450f-9958-876d3834eb5a.png)

If you check the number of NaN normals you get 15.

```julia
m = CairoMakie.surface2mesh(to_value.(p.converted)...)
isnan.(m.normals) |> sum
```

If you mark them you get this (not showing position with NaN values in z)

![Screenshot from 2023-01-17 16-18-05](https://user-images.githubusercontent.com/10947937/212937536-bb14ffbf-5f66-4c08-9bb0-e3aab80a41e1.png)

The bottom right corner is probably not in a triangle with a NaN position, so the normal is clean. The two on left - no idea... 

With
```julia
function nan_aware_orthogonal_vector(v1, v2, v3)
    (isnan(v1) || isnan(v2) || isnan(v3)) && return Vec3f(0)
    Vec3f(cross(v2 - v1, v3 - v1))
end
```
you get

![Screenshot from 2023-01-17 16-26-41](https://user-images.githubusercontent.com/10947937/212939416-c8b65e63-4eef-4f1d-a15e-87bbf97d29c0.png)

in CairoMakie (with NaN normals still marked, new zs/ns)

Looking at this in more detail also brings up two more questions:
- Why is GLMakie extending the surface past the vertex positions around NaN?
- Should CairoMakie be iterating quad faces to avoid getting triangles in corners? (I think we could just filter faces containing a NaN coordinate for this in `surface2mesh`)

Reply:
I had tried quad faces for surface earlier, but file sizes ballooned so it turned out to be counterproductive.